yaptide/celery/tasks.py

```diff
@@ -2,210 +2,197 @@
 
+class SimulationStats():
+    """Class holding simulation statistics"""
 
-def write_input_files(param_dict: dict, raw_input_dict: dict, output_dir: Path):
-    """
-    Function used to write input files to output directory.
-    Returns dictionary with filenames as keys and their content as values
-    """
-    if "beam.dat" not in raw_input_dict:
-        conv_parser = get_parser_from_str(param_dict["sim_type"])
-        return run_parser(parser=conv_parser, input_data=raw_input_dict, output_dir=output_dir)
-
-    for key, file in raw_input_dict.items():
-        with open(Path(output_dir, key), "w") as writer:
-            writer.write(file)
-    return raw_input_dict
+    def __init__(self, ntasks: int, parent, parent_id: str):
+        self.lock = Lock()
+        self.tasks_status = {}
+        self.parent = parent
+        self.parent_id = parent_id
+        for i in range(ntasks):
+            self.tasks_status[str(i+1)] = {
+                "task_id": i+1,
+                "task_state": SimulationModel.JobStatus.PENDING.value
+            }
+        parent_state = AsyncResult(parent_id).state
+        parent_meta = AsyncResult(parent_id).info
+        parent_meta["job_tasks_status"] = self.to_list()
+        self.parent.update_state(task_id=self.parent_id, state=parent_state, meta=parent_meta)
+
+    def update(self, task_id: str, up_dict: dict, final: bool = False):
+        """Method updating simulation statistics"""
+        self.lock.acquire()
+        try:
+            for key, value in up_dict.items():
+                self.tasks_status[task_id][key] = value
+            if final:
+                self.tasks_status[task_id]["simulated_primaries"] = self.tasks_status[task_id]["requested_primaries"]
+                self.tasks_status[task_id].pop("estimated_time", None)
+            parent_state = AsyncResult(self.parent_id).state
+            parent_meta = AsyncResult(self.parent_id).info
+            parent_meta["job_tasks_status"] = self.to_list()
+            self.parent.update_state(task_id=self.parent_id, state=parent_state, meta=parent_meta)
+        finally:
+            self.lock.release()
+
+    def to_list(self) -> list:
+        """To list"""
+        return [value for _, value in self.tasks_status.items()]
+
+
+class SharedResourcesManager(BaseManager):
+    """Shared objects manager for multiprocessing"""
+
+
+def log_generator(thefile):
+    """Generator function for monitoring purpose"""
+    while True:
+        line = thefile.readline()
+        if not line:
+            time.sleep(1)
+            continue
+        yield line
+
+
+def read_file(stats: SimulationStats, filepath: Path, task_id: int):
+    """Monitoring function"""
+    run_match = r"\bPrimary particle no.\s*\d*\s*ETR:\s*\d*\s*hour.*\d*\s*minute.*\d*\s*second.*\b"
+    complete_match = r"\bRun time:\s*\d*\s*hour.*\d*\s*minute.*\d*\s*second.*\b"
+    requested_match = r"\bRequested number of primaries NSTAT"
+
+    while True:
```

Comment:
what if the file will never be created ? i.e. shieldhit is crashing before saving logs ?
maybe add some logic with attempt counter.
we can assume the file is created no longer that 30 seconds from the start.
Therefore add a logic to enter except clause only max 30 times

Reply:
Done